### PR TITLE
[improvement](binlog)Support inverted index in CCR

### PR DESCRIPTION
**Pull request by qidaye**
Support inverted index in several occasions:
1. Creating binlog files
2. Downloading binlog files
3. During engine clone tasks
4. Binlog gc

## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by dataroaring** [comment]
please add cases.

**Comment by qidaye** [comment]
run buildall

File: be/src/olap/rowset/beta_rowset.cpp [hunk 1]
```diff
@@ -470,6 +470,27 @@ Status BetaRowset::add_to_binlog() {
...
+            const auto* index_info = tablet_schema()->get_inverted_index(column);
+            auto index_id = index_info->index_id();
+            for (int i = 0; i < segments_num; ++i) {
+                auto index_file = InvertedIndexDescriptor::inverted_index_file_path(
+                        _rowset_dir, rowset_id(), i, index_id, index_info->get_index_suffix());
+                auto binlog_index_file = (std::filesystem::path(binlog_dir) /
+                                          std::filesystem::path(index_file).filename())
+                                                 .string();
+                VLOG_DEBUG << "link " << index_file << " to " << binlog_index_file;
+                if (!local_fs->link_file(index_file, binlog_index_file).ok()) {
```
Comment (xiaokang) [inline be/src/olap/rowset/beta_rowset.cpp:485]: add linked file gc if some failed just like link_files_to()

File: be/src/olap/snapshot_manager.cpp [hunk 2]
```diff
@@ -635,6 +636,44 @@ Status SnapshotManager::_create_snapshot_files(const TabletSharedPtr& ref_tablet
...
+                    for (int i = 0; i < num_segments; ++i) {
+                        auto index_file =
+                                ref_tablet->get_segment_index_filepath(rowset_id, i, index_id);
+                        auto snapshot_segment_index_file_path =
+                                fmt::format("{}/{}_{}_{}.binlog-index", schema_full_path, rowset_id,
+                                            i, index_id);
+
+                        VLOG_DEBUG << "link " << index_file << " to "
+                                   << snapshot_segment_index_file_path;
+                        res = io::global_local_filesystem()->link_file(
```
Comment (xiaokang) [inline be/src/olap/snapshot_manager.cpp:667]: linked files gc on failure

File: be/src/olap/rowset/beta_rowset.cpp [hunk 3]
```diff
@@ -470,6 +470,27 @@ Status BetaRowset::add_to_binlog() {
         }
     }
 
+    // link inverted index files
+    for (const auto& column : tablet_schema()->columns()) {
+        if (tablet_schema()->has_inverted_index(column)) {
+            const auto* index_info = tablet_schema()->get_inverted_index(column);
```
Comment (xiaokang) [inline be/src/olap/rowset/beta_rowset.cpp:476]: consistent name index_meta

File: be/src/olap/snapshot_manager.cpp (hunk 2 above)
Comment (xiaokang) [reply be/src/olap/snapshot_manager.cpp:667]: why are there two link file operations for CCR?

File: be/src/olap/tablet.cpp [hunk 4]
```diff
@@ -2657,6 +2657,19 @@ std::string Tablet::get_segment_filepath(std::string_view rowset_id, int64_t seg
     return fmt::format("{}/_binlog/{}_{}.dat", _tablet_path, rowset_id, segment_index);
 }
 
+std::string Tablet::get_segment_index_filepath(std::string_view rowset_id,
+                                               std::string_view segment_index,
+                                               std::string_view index_index) const {
+    return fmt::format("{}/_binlog/{}_{}_{}.idx", _tablet_path, rowset_id, segment_index,
```
Comment (xiaokang) [inline be/src/olap/tablet.cpp:2663]: do not forget to change it after index file format v2 refactor

File: be/src/olap/tablet.cpp [hunk 5]
```diff
@@ -2657,6 +2657,19 @@ std::string Tablet::get_segment_filepath(std::string_view rowset_id, int64_t seg
     return fmt::format("{}/_binlog/{}_{}.dat", _tablet_path, rowset_id, segment_index);
 }
 
+std::string Tablet::get_segment_index_filepath(std::string_view rowset_id,
+                                               std::string_view segment_index,
+                                               std::string_view index_index) const {
```
Comment (xiaokang) [inline be/src/olap/tablet.cpp:2662]: Do you mean index_id by index_index ?

File: be/src/service/backend_service.cpp [hunk 6]
```diff
@@ -286,8 +286,121 @@ void _ingest_binlog(StorageEngine& engine, IngestBinlogArg* arg) {
...
+                             << ", get_segment_index_file_url=" << get_segment_index_file_url
+                             << ", index_file_size=" << segment_index_file_size
+                             << ", local_index_file_size=" << local_index_file_size;
+                return Status::InternalError("downloaded index file size is not equal");
+            }
+            return io::global_local_filesystem()->permission(local_segment_index_path,
+                                                             io::LocalFileSystem::PERMS_OWNER_RW);
+        };
+
+        status = HttpClient::execute_with_retry(max_retry, 1, get_segment_index_file_cb);
```
Comment (xiaokang) [inline be/src/service/backend_service.cpp:393]: Is there gc logic for success files on failure?

**Comment by qidaye** [comment]
> please add cases.

Case is here: https://github.com/selectdb/ccr-syncer/pull/48

File: be/src/olap/rowset/beta_rowset.cpp (hunk 1 above)
Comment (qidaye) [reply be/src/olap/rowset/beta_rowset.cpp]: done

File: be/src/service/backend_service.cpp (hunk 6 above)
Comment (qidaye) [reply be/src/service/backend_service.cpp:400]: Added.

**Review by xiaokang — DISMISSED** [review]
LGTM

**Comment by qidaye** [comment]
run buildall

**Comment by qidaye** [comment]
run buildall

**Review by JackDrogon — APPROVED** [review]
LGTM

**Review by xiaokang — APPROVED** [review]
LGTM

**Pull request merged**